sdk/python/feast/infra/online_stores/snowflake.py

```diff
@@ -0,0 +1,230 @@
+            if progress:
+                progress(1)
+
+        if dfs:
+            agg_df = pd.concat(dfs)
+
+            with get_snowflake_conn(config.online_store, autocommit=False) as conn:
+                write_pandas_binary(conn, agg_df, f"{config.project}_{table.name}")
+
+                query = f"""
+                    INSERT OVERWRITE INTO "{config.online_store.database}"."{config.online_store.schema_}"."{config.project}_{table.name}"
+                        SELECT
+                            "entity_feature_key",
+                            "entity_key",
+                            "feature_name",
+                            "value",
+                            "event_ts",
+                            "created_ts"
+                        FROM
+                          (SELECT
+                              *,
+                              ROW_NUMBER() OVER(PARTITION BY "entity_key","feature_name" ORDER BY "event_ts" DESC, "created_ts" DESC) AS "_feast_row"
+                          FROM
+                              "{config.online_store.database}"."{config.online_store.schema_}"."{config.project}_{table.name}")
+                        WHERE
+                            "_feast_row" = 1;
+                """
+
+                conn.cursor().execute(query)
+
+        return None
+
+    @log_exceptions_and_usage(online_store="snowflake")
+    def online_read(
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        entity_keys: List[EntityKeyProto],
+        requested_features: List[str],
+    ) -> List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]]:
+        assert isinstance(config.online_store, SnowflakeOnlineStoreConfig)
+
+        result: List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]] = []
+
+        with get_snowflake_conn(config.online_store) as conn:
+
+            df = (
+                conn.cursor()
+                .execute(
+                    f"""
+                SELECT
+                    "entity_key", "feature_name", "value", "event_ts"
+                FROM
+                    "{config.online_store.database}"."{config.online_store.schema_}"."{config.project}_{table.name}"
+                WHERE
+                    "entity_feature_key" IN ({','.join([('TO_BINARY('+hexlify(serialize_entity_key(combo[0])+bytes(combo[1], encoding='utf-8')).__str__()[1:]+")") for combo in itertools.product(entity_keys,requested_features)])})
+            """,
+                )
+                .fetch_pandas_all()
+            )
+
+        for entity_key in entity_keys:
+            entity_key_bin = serialize_entity_key(entity_key)
+            res = {}
+            res_ts = None
+            for index, row in df[df["entity_key"] == entity_key_bin].iterrows():
+                val = ValueProto()
+                val.ParseFromString(row["value"])
+                res[row["feature_name"]] = val
+                res_ts = row["event_ts"].to_pydatetime()
```

Comment:
i believe our new convention to prevent issues (see this https://github.com/feast-dev/feast/issues/2906) is to make sure that all timestamps are timezone aware.

Reply:
I see you added a to_naive_utc, but probably should check here in the fetch call too

Reply:
@kevjumba I guess why doesnt this exist in other online stores? I would rather make sure in the offline materialization query that this check happens then have to spend the cpu time on the online side

Reply:
I think we're in the process of adding. But yeah this lg to me